File: src/inspectah/architect/analyzer.py
"""Cross-fleet analyzer for layer topology decomposition."""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field


@dataclass
class FleetInput:
    """Simplified fleet data for analysis."""

    name: str
    packages: list[str]
    configs: list[str]
    host_count: int = 0
    base_image: str = ""
    unavailable_packages: list[str] = field(default_factory=list)
    direct_install_packages: list[str] = field(default_factory=list)
    unverifiable_packages: list[str] = field(default_factory=list)
    preflight_status: str = "skipped"


@dataclass
class FleetInfo:
    """Fleet metadata for the topology."""

    name: str
    host_count: int
    total_packages: int


@dataclass
class Layer:
    """A layer in the image topology."""

    name: str
    parent: str | None
    packages: list[str] = field(default_factory=list)
    configs: list[str] = field(default_factory=list)
    fleets: list[str] = field(default_factory=list)
    fan_out: int = 0
    turbulence: float = 0.0

    def _recalc_turbulence(self) -> None:
        raw = self.fan_out * (len(self.packages) / 50.0)
        if self.parent is not None:  # non-base layers get floor of 1.0
            self.turbulence = max(1.0, raw)
        else:
            self.turbulence = raw


@dataclass
class LayerTopology:
    """Complete layer topology with move and export support."""

    layers: list[Layer] = field(default_factory=list)
    fleets: list[FleetInfo] = field(default_factory=list)

    def get_layer(self, name: str) -> Layer | None:
        for layer in self.layers:
            if layer.name == name:
                return layer
        return None

    def move_package(self, package: str, from_layer: str, to_layer: str) -> None:
        """Move a package between layers.

        Standard move: remove from from_layer, add to to_layer.
        Special case — moving FROM base: package is broadcast to ALL
        derived layers (every fleet still needs it). The to_layer parameter
        identifies the user's chosen target; broadcast is a side effect.
        """
        src = self.get_layer(from_layer)
        dst = self.get_layer(to_layer)
        if src is None:
            raise ValueError(f"Layer {from_layer!r} not found")
        if dst is None:
            raise ValueError(f"Layer {to_layer!r} not found")
        if package not in src.packages:
            raise ValueError(f"Package {package!r} not found in layer {from_layer!r}")

        src.packages.remove(package)

        if from_layer == "base":
            # Broadcast to ALL derived layers
            for layer in self.layers:
                if layer.parent is not None and package not in layer.packages:
                    layer.packages.append(package)
        else:
            if package not in dst.packages:
                dst.packages.append(package)

        self._recalc_all()

    def copy_package(self, package: str, from_layer: str, to_layer: str) -> None:
        """Copy a package to another layer without removing from source."""
        src = self.get_layer(from_layer)
        dst = self.get_layer(to_layer)
        if src is None:
            raise ValueError(f"Layer {from_layer!r} not found")
        if dst is None:
            raise ValueError(f"Layer {to_layer!r} not found")
        if package not in src.packages:
            raise ValueError(f"Package {package!r} not found in layer {from_layer!r}")
        if package not in dst.packages:
            dst.packages.append(package)
        self._recalc_all()

    def _recalc_all(self) -> None:
        base = self.get_layer("base")
        if base is not None:
            base.fan_out = sum(1 for l in self.layers if l.parent == "base")
            base._recalc_turbulence()
        for layer in self.layers:
            if layer.parent is not None:
                layer.fan_out = sum(1 for l in self.layers if l.parent == layer.name)
                layer._recalc_turbulence()

    def to_dict(self) -> dict:
        return {
            "layers": [
                {
                    "name": l.name,
                    "parent": l.parent,
                    "packages": l.packages,
                    "configs": l.configs,
                    "fleets": l.fleets,
                    "fan_out": l.fan_out,
                    "turbulence": round(l.turbulence, 1),
                }
                for l in self.layers
            ],
            "fleets": [
                {"name": f.name, "host_count": f.host_count, "total_packages": f.total_packages}
                for f in self.fleets
            ],
        }


def analyze_fleets(fleets: list[FleetInput]) -> LayerTopology:
    """Analyze multiple fleets and produce a layer topology.

    Uses 100% cross-fleet prevalence heuristic: packages in ALL fleets -> base.
    Remaining packages stay in their fleet's derived layer.
    Configs always stay with their original fleet (not decomposed).
    Single fleet -> no base extraction (fleet becomes the only layer).
    """
    fleet_names = [f.name for f in fleets]
    fleet_infos = [
        FleetInfo(name=f.name, host_count=f.host_count, total_packages=len(f.packages))
        for f in fleets
    ]

    if len(fleets) == 1:
        f = fleets[0]
        layer = Layer(
            name=f.name,
            parent=None,
            packages=list(f.packages),
            configs=list(f.configs),
            fleets=[f.name],
        )
        layer._recalc_turbulence()
        return LayerTopology(layers=[layer], fleets=fleet_infos)

    # Build cross-fleet package index
    pkg_to_fleets: dict[str, set[str]] = defaultdict(set)
    for f in fleets:
        for pkg in f.packages:
            pkg_to_fleets[pkg].add(f.name)

    all_fleet_names = set(fleet_names)
    base_packages = sorted(pkg for pkg, f_set in pkg_to_fleets.items() if f_set == all_fleet_names)

    # Build layers
    base = Layer(name="base", parent=None, packages=base_packages, fleets=fleet_names)

    derived_layers = []
    for f in fleets:
        derived_packages = sorted(pkg for pkg in f.packages if pkg not in base_packages)
        derived = Layer(
            name=f.name,
            parent="base",
            packages=derived_packages,
            configs=list(f.configs),
            fleets=[f.name],
        )
        derived_layers.append(derived)

    layers = [base] + derived_layers
    topo = LayerTopology(layers=layers, fleets=fleet_infos)
    topo._recalc_all()
    return topo

File: cmd/inspectah/internal/architect/analyzer.go
// Package architect provides cross-fleet layer topology decomposition.
//
// It analyzes multiple fleet merge outputs to identify shared (base) and
// role-specific (derived) package layers, producing a LayerTopology that
// can be exported as Containerfiles.
package architect

import (
	"fmt"
	"math"
	"sort"
)

// FleetInput is simplified fleet data ready for analysis.
type FleetInput struct {
	Name                   string
	Packages               []string
	Configs                []string
	HostCount              int
	BaseImage              string
	UnavailablePackages    []string
	DirectInstallPackages  []string
	UnverifiablePackages   []string
	PreflightStatus        string
}

// FleetInfo holds fleet metadata for the topology.
type FleetInfo struct {
	Name          string `json:"name"`
	HostCount     int    `json:"host_count"`
	TotalPackages int    `json:"total_packages"`
}

// Layer represents a layer in the image topology.
type Layer struct {
	Name       string   `json:"name"`
	Parent     *string  `json:"parent"`
	Packages   []string `json:"packages"`
	Configs    []string `json:"configs"`
	Fleets     []string `json:"fleets"`
	FanOut     int      `json:"fan_out"`
	Turbulence float64  `json:"turbulence"`
}

// recalcTurbulence recomputes the turbulence score for a layer.
func (l *Layer) recalcTurbulence() {
	raw := float64(l.FanOut) * (float64(len(l.Packages)) / 50.0)
	if l.Parent != nil { // non-base layers get floor of 1.0
		l.Turbulence = math.Max(1.0, raw)
	} else {
		l.Turbulence = raw
	}
}

// LayerTopology is the complete layer topology with manipulation support.
type LayerTopology struct {
	Layers []Layer     `json:"layers"`
	Fleets []FleetInfo `json:"fleets"`
}

// GetLayer returns the layer with the given name, or nil if not found.
func (t *LayerTopology) GetLayer(name string) *Layer {
	for i := range t.Layers {
		if t.Layers[i].Name == name {
			return &t.Layers[i]
		}
	}
	return nil
}

// MovePackage moves a package between layers.
//
// Standard move: remove from src, add to dst.
// Special case -- moving FROM base: package is broadcast to ALL derived
// layers (every fleet still needs it). The toLayer parameter identifies
// the user's chosen target; broadcast is a side effect.
func (t *LayerTopology) MovePackage(pkg, fromLayer, toLayer string) error {
	src := t.GetLayer(fromLayer)
	if src == nil {
		return fmt.Errorf("layer %q not found", fromLayer)
	}
	dst := t.GetLayer(toLayer)
	if dst == nil {
		return fmt.Errorf("layer %q not found", toLayer)
	}
	if !containsString(src.Packages, pkg) {
		return fmt.Errorf("package %q not found in layer %q", pkg, fromLayer)
	}

	src.Packages = removeString(src.Packages, pkg)

	if fromLayer == "base" {
		// Broadcast to ALL derived layers
		for i := range t.Layers {
			if t.Layers[i].Parent != nil && !containsString(t.Layers[i].Packages, pkg) {
				t.Layers[i].Packages = append(t.Layers[i].Packages, pkg)
			}
		}
	} else {
		if !containsString(dst.Packages, pkg) {
			dst.Packages = append(dst.Packages, pkg)
		}
	}

	t.recalcAll()
	return nil
}

// CopyPackage copies a package to another layer without removing from source.
func (t *LayerTopology) CopyPackage(pkg, fromLayer, toLayer string) error {
	src := t.GetLayer(fromLayer)
	if src == nil {
		return fmt.Errorf("layer %q not found", fromLayer)
	}
	dst := t.GetLayer(toLayer)
	if dst == nil {
		return fmt.Errorf("layer %q not found", toLayer)
	}
	if !containsString(src.Packages, pkg) {
		return fmt.Errorf("package %q not found in layer %q", pkg, fromLayer)
	}
	if !containsString(dst.Packages, pkg) {
		dst.Packages = append(dst.Packages, pkg)
	}
	t.recalcAll()
	return nil
}

// ToDict returns the topology as a serializable map.
func (t *LayerTopology) ToDict() map[string]interface{} {
	layers := make([]map[string]interface{}, len(t.Layers))
	for i, l := range t.Layers {
		layers[i] = map[string]interface{}{
			"name":       l.Name,
			"parent":     l.Parent,
			"packages":   l.Packages,
			"configs":    l.Configs,
			"fleets":     l.Fleets,
			"fan_out":    l.FanOut,
			"turbulence": math.Round(l.Turbulence*10) / 10,
		}
	}
	fleets := make([]map[string]interface{}, len(t.Fleets))
	for i, f := range t.Fleets {
		fleets[i] = map[string]interface{}{
			"name":           f.Name,
			"host_count":     f.HostCount,
			"total_packages": f.TotalPackages,
		}
	}
	return map[string]interface{}{
		"layers": layers,
		"fleets": fleets,
	}
}

// recalcAll recomputes fan_out and turbulence for all layers.
func (t *LayerTopology) recalcAll() {
	base := t.GetLayer("base")
	for i := range t.Layers {
		l := &t.Layers[i]
		if base != nil && l.Parent == nil && l.Name == "base" {
			// Base fan_out = number of derived layers
			l.FanOut = 0
			for _, other := range t.Layers {
				if other.Parent != nil {
					l.FanOut++
				}
			}
		} else if l.Parent != nil {
			l.FanOut = 1
		}
		l.recalcTurbulence()
	}
}

// AnalyzeFleets analyzes multiple fleets and produces a layer topology.
//
// Uses 100% cross-fleet prevalence heuristic: packages in ALL fleets go
// to base. Remaining packages stay in their fleet's derived layer.
// Configs always stay with their original fleet (not decomposed).
// Single fleet -> no base extraction (fleet becomes the only layer).
func AnalyzeFleets(fleets []FleetInput) *LayerTopology {
	fleetInfos := make([]FleetInfo, len(fleets))
	for i, f := range fleets {
		fleetInfos[i] = FleetInfo{
			Name:          f.Name,
			HostCount:     f.HostCount,
			TotalPackages: len(f.Packages),
		}
	}

	if len(fleets) == 1 {
		f := fleets[0]
		layer := Layer{
			Name:     f.Name,
			Parent:   nil,
			Packages: copyStrings(f.Packages),
			Configs:  copyStrings(f.Configs),
			Fleets:   []string{f.Name},
		}
		layer.recalcTurbulence()
		topo := &LayerTopology{
			Layers: []Layer{layer},
			Fleets: fleetInfos,
		}
		return topo
	}

	// Build cross-fleet package index
	fleetNames := make([]string, len(fleets))
	for i, f := range fleets {
		fleetNames[i] = f.Name
	}
	allFleetSet := make(map[string]bool, len(fleetNames))
	for _, n := range fleetNames {
		allFleetSet[n] = true
	}

	pkgToFleets := make(map[string]map[string]bool)
	for _, f := range fleets {
		for _, pkg := range f.Packages {
			if pkgToFleets[pkg] == nil {
				pkgToFleets[pkg] = make(map[string]bool)
			}
			pkgToFleets[pkg][f.Name] = true
		}
	}

	// Packages in ALL fleets -> base
	var basePackages []string
	for pkg, fSet := range pkgToFleets {
		if len(fSet) == len(allFleetSet) {
			allMatch := true
			for n := range allFleetSet {
				if !fSet[n] {
					allMatch = false
					break
				}
			}
			if allMatch {
				basePackages = append(basePackages, pkg)
			}
		}
	}
	sort.Strings(basePackages)

	basePkgSet := make(map[string]bool, len(basePackages))
	for _, p := range basePackages {
		basePkgSet[p] = true
	}

	// Build layers
	base := Layer{
		Name:     "base",
		Parent:   nil,
		Packages: basePackages,
		Configs:  nil,
		Fleets:   copyStrings(fleetNames),
	}

	derivedLayers := make([]Layer, len(fleets))
	parentName := "base"
	for i, f := range fleets {
		var derivedPkgs []string
		for _, pkg := range f.Packages {
			if !basePkgSet[pkg] {
				derivedPkgs = append(derivedPkgs, pkg)
			}
		}
		sort.Strings(derivedPkgs)
		derivedLayers[i] = Layer{
			Name:     f.Name,
			Parent:   &parentName,
			Packages: derivedPkgs,
			Configs:  copyStrings(f.Configs),
			Fleets:   []string{f.Name},
		}
	}

	layers := make([]Layer, 0, 1+len(derivedLayers))
	layers = append(layers, base)
	layers = append(layers, derivedLayers...)

	topo := &LayerTopology{
		Layers: layers,
		Fleets: fleetInfos,
	}
	topo.recalcAll()
	return topo
}

// --- helpers ---

func containsString(ss []string, s string) bool {
	for _, v := range ss {
		if v == s {
			return true
		}
	}
	return false
}

func removeString(ss []string, s string) []string {
	result := make([]string, 0, len(ss))
	for _, v := range ss {
		if v != s {
			result = append(result, v)
		}
	}
	return result
}

func copyStrings(ss []string) []string {
	if ss == nil {
		return nil
	}
	out := make([]string, len(ss))
	copy(out, ss)
	return out
}
